xbmc/filesystem/SpecialProtocol.cpp

```diff
@@ -28,12 +28,402 @@
+
+std::string CSpecialProtocol::GetHomePath() 
+{
+  std::string path;
+#ifdef TARGET_WINDOWS
+  path = CWIN32Util::GetProfilePath();
+#endif
+#if defined(TARGET_POSIX) && !defined(TARGET_DARWIN)
+  if (g_application.PlatformDirectoriesEnabled()) 
+  {
+    std::string userHome;
+    std::string appName = CCompileInfo::GetAppName();
+    std::string dotLowerAppName = "." + appName;
+    StringUtils::ToLower(dotLowerAppName);
+    // REVIEW: Make a get function for userHome calc
+    if (getenv("HOME"))
+      userHome = getenv("HOME");
+    else
+      userHome = "/root";
+    path = userHome + "/" + dotLowerAppName;
+  }
+  else
+  {
+    path = URIUtils::AddFileToFolder(GetXBMCPath(), "portable_data");
+  }
+#endif
+#if defined(TARGET_DARWIN)
+  if (g_application.PlatformDirectoriesEnabled())
+  {
+    // REVIEW: Make a get function for userHome calc
+    std::string userHome;
+    if (getenv("HOME"))
+      userHome = getenv("HOME");
+    else
+      userHome = "/root";
+#if defined(TARGET_DARWIN_IOS)
+    path = userHome + "/" + CDarwinUtils::GetAppRootFolder() + "/" + CCompileInfo::GetAppName();
+#else
+    path = userHome + "/Library/Application Support/" + CCompileInfo::GetAppName();
+#endif
+  }
+  else
+  {
+    path = URIUtils::AddFileToFolder(GetXBMCPath(), "portable_data");
+  }
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetHomePath(const std::string &fileName) 
+{
+  return URIUtils::AddFileToFolder(GetHomePath(), fileName);
+}
+
+std::string CSpecialProtocol::GetEnvHomePath()
+{
+  std::string path;
+#if defined(TARGET_POSIX)
+  path =  getenv("HOME");
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetEnvHomePath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetEnvHomePath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetTmpPath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetTmpPath(), fileName);
+}
+
+std::string CSpecialProtocol::GetTmpPath() 
+{
+  std::string path;
+#ifdef TARGET_WINDOWS
+  path = URIUtils::AddFileToFolder(CWIN32Util::GetProfilePath(), "cache");
+#endif
+#if defined(TARGET_POSIX) && !defined(TARGET_DARWIN)
+  const char* envAppTemp = "KODI_TEMP";
+  // REVIEW: Make a get function for userHome calc
+  std::string userHome;
+  if (getenv("HOME"))
+    userHome = getenv("HOME");
+  else
+    userHome = "/root";
+
+  if (g_application.PlatformDirectoriesEnabled())
+  {
+    std::string strTempPath = userHome;
+    std::string appName = CCompileInfo::GetAppName();
+    // REVIEW: Make a get function for dotLowerAppName calc
+    std::string dotLowerAppName = "." + appName;
+    StringUtils::ToLower(dotLowerAppName);
+    strTempPath = URIUtils::AddFileToFolder(strTempPath, dotLowerAppName + "/temp");
+    if (getenv(envAppTemp))
+      strTempPath = getenv(envAppTemp);
+    path = strTempPath;
+  }
+  else
+  {
+    std::string strTempPath = GetXBMCPath();
+    strTempPath = URIUtils::AddFileToFolder(strTempPath, "portable_data/temp");
```

Comment:
`strTempPath = URIUtils::AddFileToFolder(strTempPath, "portable_data", "temp");`

Reply:
And even better is to move hard-coded strings to the top of the .cpp files using #defines or constants

Reply:
Moving the constants is a good idea as well. I think this can go in even without that as it's his/her first PR and some cleanup can be done in a second pr. Now that I think about it, same goes for my comment about the helper method. It can be refactored later.

I'm feeling bad about this taking so long. I need to go over my previous comments and see what what affects functionality and what's more cosmetic that can be bumped.